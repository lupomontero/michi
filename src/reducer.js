const initialState = {
  currPlayer: 'X',
  winner: null,
  winnerCombination: null,
  isDraw: null,
  matrix: [
    [ null, null, null ],
    [ null, null, null ],
    [ null, null, null ],
  ],
};


// chequea eje x
const isWinnerX = (currPlayer, payload, matrix) => matrix.reduce(
  (memo, row) => row[payload.y] === currPlayer && memo,
  'x',
);

// chequea eje y
const isWinnerY = (currPlayer, payload, matrix) => matrix[payload.x].reduce(
  (memo, value) => value === currPlayer && memo,
  'y',
);

// chequea diagonal (0,0) -> (2, 2)
const isWinnerDiagonalDown = (currPlayer, matrix) => matrix.reduce(
  (memo, row, x) => row.reduce(
    (memo, value, y) => (x === y && value !== currPlayer) ? false : memo,
    memo,
  ),
  'dd',
);

// chequea diagonal (0,2) -> (2, 0)
const isWinnerDiagonalUp = (currPlayer, matrix) => matrix.reduce(
  (memo, row, x) => row.reduce(
    (memo, value, y) =>
      (x === (row.length - 1) - y && value !== currPlayer) ? false : memo,
    memo,
  ),
  'du',
);

const isWinner = (currPlayer, payload, matrix) =>
  isWinnerX(currPlayer, payload, matrix)
    || isWinnerY(currPlayer, payload, matrix)
    || isWinnerDiagonalDown(currPlayer, matrix)
    || isWinnerDiagonalUp(currPlayer, matrix);

const getWinnerCombination = (currPlayer, payload, matrix) => {
  const winnerAxis = isWinner(currPlayer, payload, matrix);
  return (winnerAxis === 'x' && matrix.map((row, x) => [x, payload.y]))
    || (winnerAxis === 'y' && matrix[payload.x].map((row, y) => [payload.x, y]))
    || (winnerAxis === 'dd' && matrix.map((row, x) => [x, x]))
    || (winnerAxis === 'du' && matrix.map((row, x) => [x, row.length - 1 - x]));
};

const isDraw = matrix => matrix.reduce(
  (memo, row) => row.reduce((memo, value) => memo && !!value, memo),
  true,
);


export default (state = initialState, { type, payload }) => {
  if (type === 'RESET') {
    return initialState;
  }

  if (type !== 'MOVE') {
    return state;
  }

  if (state.matrix[payload.x][payload.y]) {
    return state;
  }

  const matrix = state.matrix.map(
    (row, x) => row.map((value, y) => (
      (!value && x === payload.x && y === payload.y)
        ? state.currPlayer
        : value
    ))
  );

  const winnerCombination = getWinnerCombination(state.currPlayer, payload, matrix);

  return {
    currPlayer: (state.currPlayer === 'X') ? 'O' : 'X',
    winner: !!winnerCombination ? state.currPlayer : null,
    winnerCombination,
    isDraw: isDraw(matrix),
    matrix,
  };
};
