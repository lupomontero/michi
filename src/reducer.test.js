import reducer from './reducer';

describe('reducer', () => {

  it('debería ser una función', () => {
    expect(typeof reducer).toBe('function');
  });

  it('debería retornar el estado inicial cuando no recibe estado', () => {
	  expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('debería empezar con jugador X y cambiar turno después de mover', () => {
	  expect(reducer(undefined, { type: 'MOVE', payload: { x: 0, y: 0 } })).toMatchSnapshot();
  });

  it('no debería dejar marcar una casilla ya marcada', () => {
    const state1 = reducer(undefined, { type: 'MOVE', payload: { x: 1, y: 1 } });
    const state2 = reducer(state1, { type: 'MOVE', payload: { x: 1, y: 1 } });
    expect(state2).toMatchSnapshot();
  });

  it('debería marcar casilla de O en su turno', () => {
    const state1 = reducer(undefined, { type: 'MOVE', payload: { x: 0, y: 0 } });
	  expect(reducer(state1, { type: 'MOVE', payload: { x: 1, y: 0 } })).toMatchSnapshot();
  });

  it('debería resetear estado cuando acción es RESET', () => {
    const state1 = reducer(undefined, { type: 'MOVE', payload: { x: 0, y: 0 } });
    const state2 = reducer(state1, { type: 'MOVE', payload: { x: 1, y: 0 } });
    expect(reducer(state2, { type: 'RESET' })).toMatchSnapshot();
  });

  it('debería detectar combinación ganadora en eje X');

  it('debería detectar combinación ganadora en eje Y');

  it('debería detectar diagonal (0,0) -> (2,2)', () => {
    const state1 = reducer(undefined, { type: 'MOVE', payload: { x: 0, y: 0 } });
    const state2 = reducer(state1, { type: 'MOVE', payload: { x: 0, y: 1 } });
    const state3 = reducer(state2, { type: 'MOVE', payload: { x: 1, y: 1 } });
    const state4 = reducer(state3, { type: 'MOVE', payload: { x: 1, y: 0 } });
    const state5 = reducer(state4, { type: 'MOVE', payload: { x: 2, y: 2 } });
    expect(state5).toMatchSnapshot();
  });

  it('debería detectar diagonal (0,2) -> (2,0)');

  it('debería detectar empate');

});
