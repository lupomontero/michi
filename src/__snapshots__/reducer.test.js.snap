// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer debería detectar diagonal (0,0) -> (2,2) 1`] = `
Object {
  "currPlayer": "O",
  "isDraw": false,
  "matrix": Array [
    Array [
      "X",
      "O",
      null,
    ],
    Array [
      "O",
      "X",
      null,
    ],
    Array [
      null,
      null,
      "X",
    ],
  ],
  "winner": "X",
  "winnerCombination": Array [
    Array [
      0,
      0,
    ],
    Array [
      1,
      1,
    ],
    Array [
      2,
      2,
    ],
  ],
}
`;

exports[`reducer debería empezar con jugador X y cambiar turno después de mover 1`] = `
Object {
  "currPlayer": "O",
  "isDraw": false,
  "matrix": Array [
    Array [
      "X",
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "winner": null,
  "winnerCombination": false,
}
`;

exports[`reducer debería marcar casilla de O en su turno 1`] = `
Object {
  "currPlayer": "X",
  "isDraw": false,
  "matrix": Array [
    Array [
      "X",
      null,
      null,
    ],
    Array [
      "O",
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "winner": null,
  "winnerCombination": false,
}
`;

exports[`reducer debería resetear estado cuando acción es RESET 1`] = `
Object {
  "currPlayer": "X",
  "isDraw": null,
  "matrix": Array [
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "winner": null,
  "winnerCombination": null,
}
`;

exports[`reducer debería retornar el estado inicial cuando no recibe estado 1`] = `
Object {
  "currPlayer": "X",
  "isDraw": null,
  "matrix": Array [
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "winner": null,
  "winnerCombination": null,
}
`;

exports[`reducer no debería dejar marcar una casilla ya marcada 1`] = `
Object {
  "currPlayer": "O",
  "isDraw": false,
  "matrix": Array [
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      "X",
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "winner": null,
  "winnerCombination": false,
}
`;
